synapse/handlers/room_list.py

```diff
@@ -70,20 +71,24 @@ def get_local_public_room_list(self, limit=None, since_token=None,
         if search_filter:
             # We explicitly don't bother caching searches or requests for
             # appservice specific lists.
+            logger.info("Bypassing cache as search request.")
             return self._get_public_room_list(
                 limit, since_token, search_filter, network_tuple=network_tuple,
             )
 
         key = (limit, since_token, network_tuple)
         result = self.response_cache.get(key)
         if not result:
+            logger.info("No cached result, calculating one.")
             result = self.response_cache.set(
                 key,
-                self._get_public_room_list(
+                preserve_fn(self._get_public_room_list)(
                     limit, since_token, network_tuple=network_tuple
                 )
             )
-        return result
+        else:
+            logger.info("Using cached result.")
```

Comment:
hrm. slightly misleading given the result is a deferred which is (by definition) not yet complete.

possibly just "using cached deferred result" ?